quickwit/quickwit-doc-mapper/src/routing_expression/mod.rs

```diff
@@ -209,26 +284,110 @@ impl Display for InnerRoutingExpr {
 enum ExprType {
     Field,
     Composite,
+    Modulo,
 }
 
-impl RoutingExpr {
-    /// Evaluates the expression applied to the given
-    /// context and returns a u64 hash.
-    ///
-    /// Obviously this function is not perfectly injective.
-    pub fn eval_hash<Ctx: RoutingExprContext>(&self, ctx: &Ctx) -> u64 {
-        if let Some(inner) = self.inner_opt.as_ref() {
-            let mut hasher: SipHasher = self.salted_hasher;
-            inner.eval_hash(ctx, &mut hasher);
-            hasher.finish()
+mod expression_dsl {
+    use nom::bytes::complete::tag;
+    use nom::character::complete::multispace0;
+    use nom::combinator::{eof, opt};
+    use nom::error::ErrorKind;
+    use nom::multi::separated_list0;
+    use nom::sequence::{delimited, tuple};
+    use nom::{AsChar, Finish, IResult, InputTakeAtPosition};
+
+    // this is a RoutingSubExpr in our DSL.
+    #[derive(Debug, PartialEq, Eq, Clone)]
+    pub(crate) enum ExpressionAst {
+        Field(String),
+        Function { name: String, args: Vec<Argument> },
+    }
+
+    #[derive(Debug, PartialEq, Eq, Clone)]
+    pub(crate) enum Argument {
+        Expression(Vec<ExpressionAst>),
+        Number(u64),
+    }
+
+    pub(crate) fn parse_expression(expr_dsl_str: &str) -> anyhow::Result<Vec<ExpressionAst>> {
+        let (i, res) = routing_expr(expr_dsl_str)
+            .finish()
+            .map_err(|e| anyhow::anyhow!("error parsing routing expression: {e}"))?;
+        t(eof(i))?;
+
+        Ok(res)
+    }
+
+    // DSL:
+    //
+    // RoutingExpr := RoutingSubExpr [ , RoutingExpr ]
+    // RougingSubExpr := Identifier [ \( Arguments \) ]
+    // Identifier := FieldChar [ Identifier ]
+    // FieldChar := { a..z | A..Z | 0..9 | _ }
+    // Arguments := Argument [ ; Arguments ]
```

Comment:
```suggestion
    // Arguments := Argument [ ; Arguments ]
```
`;` is a weird separator for arguments. Can we make it ","?

Reply:
`,` would be ambiguous with an argument being a list of fields:
`hash_mod(tennant_id,app_id, 5)`

We could either: 
- keep the `;` as arg delimiter
- use an other delimiter than `,` for list of fields
- don't allow functions to take a list of fields, but be variadic (the example would get parsed to 3 args, annd the function would know the last is supposed to be a number, and fuse all others in a Composite)

Reply:
Parenthesis?

`hash_mod((tenant_id,app_id), 5)`